Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,222 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import GRPCCore
+import NIOCore
+
+// TODO: this is all done from client's perspective for now - work on server-side later.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+struct GRPCStreamStateMachine {
+  struct Configuration {
+    let maximumPayloadSize: Int
+    let compressor: Zlib.Compressor?
+    let decompressor: Zlib.Decompressor?
+  }
+
+  private var state: State
+  
+  enum State {
+    case clientIdleServerIdle(Configuration)
+    case clientOpenServerIdle(ClientOpenState)
+    case clientOpenServerOpen(ClientOpenState)
+    case clientOpenServerClosed(ClientOpenState)
+    case clientClosedServerIdle
+    case clientClosedServerOpen
+    case clientClosedServerClosed
+  }
+  
+  struct ClientOpenState {
+    var framer: GRPCMessageFramer
+    let deframer: NIOSingleStepByteToMessageProcessor<GRPCMessageDeframer>
+    var compressor: Zlib.Compressor?
+    
+    init(
+      maximumPayloadSize: Int,
+      compressor: Zlib.Compressor?,
+      decompressor: Zlib.Decompressor?
+    ) {
+      self.framer = GRPCMessageFramer()
+      let messageDeframer = GRPCMessageDeframer(
+        maximumPayloadSize: maximumPayloadSize,
+        decompressor: decompressor
+      )
+      self.deframer = NIOSingleStepByteToMessageProcessor(messageDeframer)
+      self.compressor = compressor
+    }
+  }
+  
+  init(configuration: Configuration) {
+    self.state = .clientIdleServerIdle(configuration)
+  }
+  
+  mutating func send(metadata: Metadata) throws {
+    // Client only sends metadata when opening.
+    try self.clientSentMetadata()
+  }
+  
+  mutating func send(message: [UInt8]) {
+    // Client sends message.
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client not yet open")
+    case .clientOpenServerIdle(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerOpen(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending a message if the server's closed.
+      ()
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Client is closed, cannot send a message")
+    }
+  }
+  
+  mutating func send(status: String, trailingMetadata: Metadata) throws {
+    // Only server does this.
```

Comment:
The client does need to send an end though.

Reply:
Right, but it's not sent in the form of a status and trailer, right? IIUC the END_STREAM flag is set at the end of the sent message(s)

Reply:
Correct, it was to highlight we still need a state transition. I saw `clientSentEnd` further but didn't think to return to and remove this comment.